File: packages/coding-agent/src/utils/shell.ts
import { existsSync } from "node:fs";
import { delimiter } from "node:path";
import { spawn, spawnSync } from "child_process";
import { getBinDir, getSettingsPath } from "../config.js";
import { SettingsManager } from "../core/settings-manager.js";

let cachedShellConfig: { shell: string; args: string[] } | null = null;

/**
 * Find bash executable on PATH (cross-platform)
 */
function findBashOnPath(): string | null {
	if (process.platform === "win32") {
		// Windows: Use 'where' and verify file exists (where can return non-existent paths)
		try {
			const result = spawnSync("where", ["bash.exe"], { encoding: "utf-8", timeout: 5000 });
			if (result.status === 0 && result.stdout) {
				const firstMatch = result.stdout.trim().split(/\r?\n/)[0];
				if (firstMatch && existsSync(firstMatch)) {
					return firstMatch;
				}
			}
		} catch {
			// Ignore errors
		}
		return null;
	}

	// Unix: Use 'which' and trust its output (handles Termux and special filesystems)
	try {
		const result = spawnSync("which", ["bash"], { encoding: "utf-8", timeout: 5000 });
		if (result.status === 0 && result.stdout) {
			const firstMatch = result.stdout.trim().split(/\r?\n/)[0];
			if (firstMatch) {
				return firstMatch;
			}
		}
	} catch {
		// Ignore errors
	}
	return null;
}

/**
 * Get shell configuration based on platform.
 * Resolution order:
 * 1. User-specified shellPath in settings.json
 * 2. On Windows: Git Bash in known locations, then bash on PATH
 * 3. On Unix: /bin/bash, then bash on PATH, then fallback to sh
 */
export function getShellConfig(): { shell: string; args: string[] } {
	if (cachedShellConfig) {
		return cachedShellConfig;
	}

	const settings = SettingsManager.create();
	const customShellPath = settings.getShellPath();

	// 1. Check user-specified shell path
	if (customShellPath) {
		if (existsSync(customShellPath)) {
			cachedShellConfig = { shell: customShellPath, args: ["-c"] };
			return cachedShellConfig;
		}
		throw new Error(
			`Custom shell path not found: ${customShellPath}\nPlease update shellPath in ${getSettingsPath()}`,
		);
	}

	if (process.platform === "win32") {
		// 2. Try Git Bash in known locations
		const paths: string[] = [];
		const programFiles = process.env.ProgramFiles;
		if (programFiles) {
			paths.push(`${programFiles}\\Git\\bin\\bash.exe`);
		}
		const programFilesX86 = process.env["ProgramFiles(x86)"];
		if (programFilesX86) {
			paths.push(`${programFilesX86}\\Git\\bin\\bash.exe`);
		}

		for (const path of paths) {
			if (existsSync(path)) {
				cachedShellConfig = { shell: path, args: ["-c"] };
				return cachedShellConfig;
			}
		}

		// 3. Fallback: search bash.exe on PATH (Cygwin, MSYS2, WSL, etc.)
		const bashOnPath = findBashOnPath();
		if (bashOnPath) {
			cachedShellConfig = { shell: bashOnPath, args: ["-c"] };
			return cachedShellConfig;
		}

		throw new Error(
			`No bash shell found. Options:\n` +
				`  1. Install Git for Windows: https://git-scm.com/download/win\n` +
				`  2. Add your bash to PATH (Cygwin, MSYS2, etc.)\n` +
				`  3. Set shellPath in ${getSettingsPath()}\n\n` +
				`Searched Git Bash in:\n${paths.map((p) => `  ${p}`).join("\n")}`,
		);
	}

	// Unix: try /bin/bash, then bash on PATH, then fallback to sh
	if (existsSync("/bin/bash")) {
		cachedShellConfig = { shell: "/bin/bash", args: ["-c"] };
		return cachedShellConfig;
	}

	const bashOnPath = findBashOnPath();
	if (bashOnPath) {
		cachedShellConfig = { shell: bashOnPath, args: ["-c"] };
		return cachedShellConfig;
	}

	cachedShellConfig = { shell: "sh", args: ["-c"] };
	return cachedShellConfig;
}

export function getShellEnv(): NodeJS.ProcessEnv {
	const binDir = getBinDir();
	const pathKey = Object.keys(process.env).find((key) => key.toLowerCase() === "path") ?? "PATH";
	const currentPath = process.env[pathKey] ?? "";
	const pathEntries = currentPath.split(delimiter).filter(Boolean);
	const hasBinDir = pathEntries.includes(binDir);
	const updatedPath = hasBinDir ? currentPath : [binDir, currentPath].filter(Boolean).join(delimiter);

	return {
		...process.env,
		[pathKey]: updatedPath,
	};
}

/**
 * Sanitize binary output for display/storage.
 * Removes characters that crash string-width or cause display issues:
 * - Control characters (except tab, newline, carriage return)
 * - Lone surrogates
 * - Unicode Format characters (crash string-width due to a bug)
 */
export function sanitizeBinaryOutput(str: unknown): string {
	if (str == null) {
		return "";
	}

	let text: string;
	if (typeof str === "string") {
		text = str;
	} else {
		try {
			text = String(str);
		} catch {
			return "";
		}
	}

	// Scan code units and accumulate safe segments without per-char arrays.
	// This keeps memory bounded and avoids Array.from() failures on very large inputs.
	let out = "";
	let segmentStart = 0;
	for (let i = 0; i < text.length; ) {
		const code = text.charCodeAt(i);
		let charLength = 1;
		let shouldDrop = false;

		if (code >= 0xd800 && code <= 0xdbff) {
			const next = i + 1 < text.length ? text.charCodeAt(i + 1) : -1;
			if (next >= 0xdc00 && next <= 0xdfff) {
				charLength = 2;
			} else {
				shouldDrop = true;
			}
		} else if (code >= 0xdc00 && code <= 0xdfff) {
			shouldDrop = true;
		} else {
			const isAllowedWhitespace = code === 0x09 || code === 0x0a || code === 0x0d;
			const isControl = code <= 0x1f;
			const isDangerousFormat = code >= 0xfff9 && code <= 0xfffb;
			shouldDrop = !isAllowedWhitespace && (isControl || isDangerousFormat);
		}

		if (shouldDrop) {
			out += text.slice(segmentStart, i);
			segmentStart = i + charLength;
		}
		i += charLength;
	}

	if (segmentStart === 0) {
		return text;
	}
	out += text.slice(segmentStart);
	return out;
}

/**
 * Detached child processes must be tracked so they can be killed on parent
 * shutdown signals (SIGHUP/SIGTERM).
 */
const trackedDetachedChildPids = new Set<number>();

export function trackDetachedChildPid(pid: number): void {
	trackedDetachedChildPids.add(pid);
}

export function untrackDetachedChildPid(pid: number): void {
	trackedDetachedChildPids.delete(pid);
}

export function killTrackedDetachedChildren(): void {
	for (const pid of trackedDetachedChildPids) {
		killProcessTree(pid);
	}
	trackedDetachedChildPids.clear();
}

/**
 * Kill a process and all its children (cross-platform)
 */
export function killProcessTree(pid: number): void {
	if (process.platform === "win32") {
		// Use taskkill on Windows to kill process tree
		try {
			spawn("taskkill", ["/F", "/T", "/PID", String(pid)], {
				stdio: "ignore",
				detached: true,
			});
		} catch {
			// Ignore errors if taskkill fails
		}
	} else {
		// Use SIGKILL on Unix/Linux/Mac
		try {
			process.kill(-pid, "SIGKILL");
		} catch {
			// Fallback to killing just the child if process group kill fails
			try {
				process.kill(pid, "SIGKILL");
			} catch {
				// Process already dead
			}
		}
	}
}

File: python/src/pi_mono/utils/shell.py
import os
import re
import shutil
import subprocess
import sys
from typing import Literal, NotRequired, TypedDict

from pi_mono.config import get_bin_dir


class ShellConfig(TypedDict):
    shell: str
    args: list[str]
    commandTransport: NotRequired[Literal["argv", "stdin"]]


def _is_legacy_wsl_bash_path(path: str) -> bool:
    normalized = path.replace("/", "\\").lower()
    return bool(re.match(r"^[a-z]:\\windows\\(?:system32|sysnative)\\bash\.exe$", normalized))


def _get_bash_shell_config(shell: str) -> ShellConfig:
    if _is_legacy_wsl_bash_path(shell):
        return {"shell": shell, "args": ["-s"], "commandTransport": "stdin"}
    return {"shell": shell, "args": ["-c"]}


def get_shell_config(custom_shell_path: str | None = None) -> ShellConfig:
    """Resolve shell configuration based on platform and optional explicit shell path."""
    if custom_shell_path:
        if os.path.exists(custom_shell_path):
            return _get_bash_shell_config(custom_shell_path)
        raise ValueError(f"Custom shell path not found: {custom_shell_path}")

    if sys.platform == "win32":
        paths = []
        program_files = os.environ.get("ProgramFiles")
        if program_files:
            paths.append(os.path.join(program_files, "Git", "bin", "bash.exe"))
        program_files_x86 = os.environ.get("ProgramFiles(x86)")
        if program_files_x86:
            paths.append(os.path.join(program_files_x86, "Git", "bin", "bash.exe"))

        for path in paths:
            if os.path.exists(path):
                return _get_bash_shell_config(path)

        bash_path = shutil.which("bash.exe") or shutil.which("bash")
        if bash_path:
            return _get_bash_shell_config(bash_path)

        raise RuntimeError(
            "No bash shell found. Options:\n"
            "  1. Install Git for Windows\n"
            "  2. Add your bash to PATH\n"
            "  3. Set shellPath in settings.json"
        )

    if os.path.exists("/bin/bash"):
        return _get_bash_shell_config("/bin/bash")

    bash_path = shutil.which("bash")
    if bash_path:
        return _get_bash_shell_config(bash_path)

    return {"shell": "sh", "args": ["-c"]}


def get_shell_env() -> dict[str, str]:
    """Get process environment with packages bin directory prepended to PATH."""
    bin_dir = str(get_bin_dir())
    env = os.environ.copy()

    # Find the PATH environment variable key in a case-insensitive manner
    path_key = "PATH"
    for key in env.keys():
        if key.upper() == "PATH":
            path_key = key
            break

    current_path = env.get(path_key, "")
    path_entries = [p for p in current_path.split(os.pathsep) if p]

    if bin_dir not in path_entries:
        path_entries.insert(0, bin_dir)
        env[path_key] = os.pathsep.join(path_entries)

    return env


def sanitize_binary_output(str_val: str) -> str:
    """Sanitize binary output for display/storage by removing formatting/control codes."""
    result = []
    for char in str_val:
        code = ord(char)
        # Allow tab, newline, carriage return
        if code in (0x09, 0x0A, 0x0D):
            result.append(char)
            continue
        # Filter control characters
        if code <= 0x1F:
            continue
        # Filter Unicode format characters
        if 0xFFF9 <= code <= 0xFFFB:
            continue
        # Filter surrogate characters
        if 0xD800 <= code <= 0xDFFF:
            continue
        result.append(char)
    return "".join(result)


# Track detached child processes
tracked_detached_child_pids: set[int] = set()


def track_detached_child_pid(pid: int) -> None:
    tracked_detached_child_pids.add(pid)


def untrack_detached_child_pid(pid: int) -> None:
    tracked_detached_child_pids.discard(pid)


def kill_tracked_detached_children() -> None:
    for pid in list(tracked_detached_child_pids):
        kill_process_tree(pid)
    tracked_detached_child_pids.clear()


def kill_process_tree(pid: int) -> None:
    """Kill a process tree (cross-platform)."""
    if sys.platform == "win32":
        try:
            subprocess.run(
                ["taskkill", "/F", "/T", "/PID", str(pid)],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                creationflags=0x08000000,  # CREATE_NO_WINDOW
            )
        except Exception:
            pass
    else:
        try:
            os.killpg(os.getpgid(pid), 9)  # SIGKILL process group
        except Exception:
            try:
                os.kill(pid, 9)  # SIGKILL single process
            except Exception:
                pass
